zebra-state/src/service/check/tests/prop.rs

```diff
@@ -0,0 +1,386 @@
+            let block1 = FinalizedBlock::from(Arc::new(block1));
+            let commit_result = state.disk.commit_finalized_direct(block1.clone(), "test");
+
+            prop_assert_eq!(Some((Height(1), block1.hash)), state.best_tip());
+            prop_assert!(commit_result.is_ok());
+
+            block1_hash = block1.hash;
+        } else {
+            let block1 = Arc::new(block1).prepare();
+            let commit_result =
+                state.validate_and_commit(block1.clone());
+
+            prop_assert_eq!(commit_result, Ok(()));
+            prop_assert_eq!(Some((Height(1), block1.hash)), state.best_tip());
+
+            block1_hash = block1.hash;
+        }
+
+        let block2 = Arc::new(block2).prepare();
+        let commit_result =
+            state.validate_and_commit(block2);
+
+        prop_assert_eq!(
+            commit_result,
+            Err(
+                DuplicateSproutNullifier {
+                    nullifier: duplicate_nullifier,
+                    in_finalized_state: duplicate_in_finalized_state,
+                }.into()
+            )
+        );
+        prop_assert_eq!(Some((Height(1), block1_hash)), state.best_tip());
+    }
+}
+
+/// Return a new `StateService` containing the mainnet genesis block.
+/// Also returns the finalized genesis block itself.
+fn new_state_with_mainnet_genesis() -> (StateService, FinalizedBlock) {
+    let genesis = zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES
+        .zcash_deserialize_into::<Arc<Block>>()
+        .expect("block should deserialize");
+
+    let mut state = StateService::new(Config::ephemeral(), Mainnet);
+
+    assert_eq!(None, state.best_tip());
+
+    let genesis = FinalizedBlock::from(genesis);
+    let commit_result = state.disk.commit_finalized_direct(genesis.clone(), "test");
+
+    assert_eq!(Some((Height(0), genesis.hash)), state.best_tip());
+    assert!(commit_result.is_ok());
+
+    (state, genesis)
+}
+
+/// Make sure the supplied nullifiers are distinct, modifying them if necessary.
+fn make_distinct_nullifiers<'joinsplit>(
+    nullifiers: impl IntoIterator<Item = &'joinsplit mut sprout::Nullifier>,
+) {
+    let nullifiers: Vec<_> = nullifiers.into_iter().collect();
+
+    if nullifiers.iter().unique().count() < nullifiers.len() {
+        let mut tweak: u8 = 0x00;
+        for nullifier in nullifiers {
+            nullifier.0[0] = tweak;
+            tweak = tweak
+                .checked_add(0x01)
+                .expect("unexpectedly large nullifier list");
+        }
+    }
+}
+
+/// Return a `Transaction::V4` containing `joinsplit_data`.
+///
+/// Other fields have empty or default values.
+fn transaction_v4_with_joinsplit_data(
+    joinsplit_data: impl Into<Option<JoinSplitData<Groth16Proof>>>,
+) -> Transaction {
+    Transaction::V4 {
+        inputs: Vec::new(),
+        outputs: Vec::new(),
+        lock_time: LockTime::min_lock_time(),
+        expiry_height: Height(0),
+        joinsplit_data: joinsplit_data.into(),
```

Comment:
TODO: set the value balances to zero, to avoid breaking future value pool tests.

This fix can happen in a later PR.

Reply:
Fixed in e947765a